File: src/main/java/ru/arvrlab/ndkcrashhandler/SignalWatcher.java
package ru.arvrlab.ndkcrashhandler;

import android.util.Log;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class SignalWatcher {
    static {
        System.loadLibrary("ndk-crash-handler");
    }

    private final Executor executor = Executors.newSingleThreadExecutor();
    public ActionAfterError actionAfterError = ()->{};

    public void startCrashHadler() {
        executor.execute(
                () -> {
                    waitForError();
                    Log.e(this.getClass().getName(), "Crash fired with error:\n" + getLastErrorMessage());
                    actionAfterError.doIt();
                }
        );
    }

    private void waitForError()  {
        boolean isError = false;
        while (!isError) {
            try { Thread.sleep(1000); } catch (InterruptedException ignored) { }
            isError = isErrorMessageExistInLog();
        }
    }

    /** Check new bytes in log
     * @return true if founded, else false
     * */
    public native boolean isErrorMessageExistInLog();
    /** Read from log file last error
     * */
    public native String getLastErrorMessage();

    public interface ActionAfterError {
        void doIt();
    }
}

File: src/main/java/ru/arvrlab/ndkcrashhandler/SignalWatcher.kt
package ru.arvrlab.ndkcrashhandler

import android.util.Log
import java.util.concurrent.Executor
import java.util.concurrent.Executors

class SignalWatcher {
    init {
        System.loadLibrary("ndk-crash-handler")
    }

    private val executor: Executor = Executors.newSingleThreadExecutor()
    var actionAfterError: Runnable? = null
    private var isWatcherEnabled = false

    /** Start watcher only once during the app lifetime.
     * @param logPath: Absolute path to log.txt
     * @param activityClass: watch over this
     * */
    fun start(logPath: String, activityClass: String) {
        if (!isWatcherEnabled) executor.execute {
            isWatcherEnabled = true
            waitForError(logPath, activityClass)
        }
    }

    /** Check log.txt for bytes each 2000ms */
    private fun waitForError(logPath: String, activityClass: String) {
        var isErrorMsgExist: Boolean
        var rescueAttempt = 0
        while (isWatcherEnabled) {
            Thread.sleep(2000)
            isErrorMsgExist = isErrorMessageExistInLog(logPath)
            Log.i(this.javaClass.name, "I'm alive.\nWatching over ${activityClass.split(".").last()} in $logPath.\nApp restarted $rescueAttempt times.")
            if (isErrorMsgExist) {
                Log.i(this.javaClass.name, "Crash fired with error:\n${getLastErrorMessage(logPath)}")
                rescueAttempt++
                actionAfterError?.run()
            }
        }
    }

    /** Stop watcher */
    fun stop(){
        isWatcherEnabled = false
        Log.i(this.javaClass.name, "Bye bye")
    }

    /** Check new bytes in log
     * @return true if founded, else false
     */
    private external fun isErrorMessageExistInLog(logPath: String): Boolean
    /** Read from log file last error
     */
    private external fun getLastErrorMessage(logPath: String): String
}